File: src/Lib/_string.py
"""string helper module"""

import re

class __loader__(object):
    pass

def formatter_field_name_split(fieldname):
    """split the argument as a field name"""
    _list=[]
    for _name in fieldname:
        _parts = _name.split('.')
        for _item in _parts:
            is_attr=False  #fix me
            if re.match('\d+', _item):
               _list.append((int(_item), is_attr))
            else:
               _list.append((_item, is_attr))

    return _list[0][0], iter(_list[1:])

def formatter_parser(*args,**kw):
    """parse the argument as a format string"""

    assert len(args)==1
    assert isinstance(args[0], str)

    _result=[]
    for _match in re.finditer("([^{]*)?(\{[^}]*\})?", args[0]):
        _pre, _fmt = _match.groups()
        if _fmt is None:
           _result.append((_pre, None, None, None))
        elif _fmt == '{}':
           _result.append((_pre, '', '', None))
        else:
           _m=re.match("\{([^!]*)!?(.*)?\}", _fmt)
           _name=_m.groups(0)
           _flags=_m.groups(1)

           _result.append((_pre, _name, _flags, None))

    return _result

File: www/src/libs/_string.js
var $module=(function($B){

var _b_ = $B.builtins

function parts(format_string){
    var result = [],
        _parts = $B.split_format(format_string) // defined in py_string.js
    for(var i = 0; i < _parts.length; i+= 2){
        result.push({pre: _parts[i], fmt: _parts[i + 1]})
    }
    return result
}

function Tuple(){
    var args = []
    for(var i=0, len=arguments.length; i < len; i++){
        args.push(arguments[i])
    }
    return _b_.tuple.$factory(args)
}

return{

    formatter_field_name_split: function(fieldname){
        // Split the argument as a field name
        var parsed = $B.parse_format(fieldname),
            first = parsed.name,
            rest = []
        if(first.match(/\d+/)){first = parseInt(first)}
        parsed.name_ext.forEach(function(ext){
            if(ext.startsWith("[")){
                var item = ext.substr(1, ext.length - 2)
                if(item.match(/\d+/)){
                    rest.push(Tuple(false, parseInt(item)))
                }else{
                    rest.push(Tuple(false, item))
                }
            }else{
                rest.push(Tuple(true, ext.substr(1)))
            }
        })
        return Tuple(first, _b_.iter(rest))
    },
    formatter_parser: function(format_string){
        // Parse the argument as a format string

        if(! _b_.isinstance(format_string, _b_.str)){
            throw _b_.ValueError.$factory("Invalid format string type: " +
                $B.class_name(format_string))
        }

        var result  = []
        parts(format_string).forEach(function(item){
            var pre = item.pre === undefined ? "" : item.pre,
                fmt = item.fmt
            if(fmt === undefined){
               result.push(Tuple(pre, _b_.None, _b_.None, _b_.None))
            }else if(fmt.string == ''){
               result.push(Tuple(pre, '', '', _b_.None))
            }else{
               result.push(Tuple(pre,
                   fmt.raw_name + fmt.name_ext.join(""),
                   fmt.raw_spec,
                   fmt.conv || _b_.None))
           }
        })
        return result
    }
}
})(__BRYTHON__)